src/commands/classified/external.rs

```diff
@@ -50,158 +49,243 @@ impl Decoder for LinesCodec {
+            name_tag,
+        ))
     }
+}
 
-    process = process.cwd(context.shell_manager.path());
+async fn run_with_stdin(
+    command: ExternalCommand,
+    context: &mut Context,
+    mut input: InputStream,
+    is_last: bool,
+) -> Result<InputStream, ShellError> {
+    let name_tag = command.name_tag.clone();
+    let home_dir = dirs::home_dir();
 
-    trace!(target: "nu::run::external", "cwd = {:?}", context.shell_manager.path());
+    let mut process = Exec::cmd(&command.name);
+    for arg in command.args.iter() {
+        // Let's also replace ~ as we shell out
+        let arg = shellexpand::tilde_with_context(arg.deref(), || home_dir.as_ref());
 
-    let mut process = match stream_next {
-        StreamNext::Last => process,
-        StreamNext::External | StreamNext::Internal => {
-            process.stdout(subprocess::Redirection::Pipe)
+        let arg_chars: Vec<_> = arg.chars().collect();
+        if (arg_chars.len() > 1 && arg_chars[0] == '"' && arg_chars[arg_chars.len() - 1] == '"')
+            || (arg_chars.len() > 1
+                && arg_chars[0] == '\''
+                && arg_chars[arg_chars.len() - 1] == '\'')
+        {
+            // quoted string
+            let new_arg: String = arg_chars[1..arg_chars.len() - 1].iter().collect();
+            process = process.arg(new_arg);
+        } else {
+            process = process.arg(arg.as_ref());
         }
-    };
+    }
 
-    trace!(target: "nu::run::external", "set up stdout pipe");
+    process = process.cwd(context.shell_manager.path());
+    trace!(target: "nu::run::external", "cwd = {:?}", context.shell_manager.path());
 
-    if let Some(stdin) = stdin {
-        process = process.stdin(stdin);
+    if !is_last {
+        process = process.stdout(subprocess::Redirection::Pipe);
+        trace!(target: "nu::run::external", "set up stdout pipe");
     }
 
+    // NOTE can't use a subprocess::make_pipe here. There seems to be no way to close stdin
+    //      after doing so, which makes certain commands block forever (e.g., cat).
+    process = process.stdin(subprocess::Redirection::Pipe);
     trace!(target: "nu::run::external", "set up stdin pipe");
+
     trace!(target: "nu::run::external", "built process {:?}", process);
 
     let popen = process.popen();
+    if let Ok(mut popen) = popen {
+        let _ = popen.detach();
+        let mut stdin_write = popen.stdin.take().unwrap();
 
-    trace!(target: "nu::run::external", "next = {:?}", stream_next);
+        // Instead of writing everything into the pipe, which may(?) overflow, we write small
+        // amounts at a time.
+        let name_tag2 = name_tag.clone();
+        let write_to_stdin_stream = async_stream! {
+            while let Some(item) = input.next().await {
+                match item.value {
+                    UntaggedValue::Primitive(Primitive::String(s)) => {
+                        stdin_write.write(s.as_bytes());
+                        stdin_write.write(b"\n");
```

Comment:
This should likely not have a \n (at least in theory). Strings that have a \n at the end automatically should be Primitive::Line

Reply:
Here's the two I'm trying to compare:

```
echo "1" "2" "3" | ^cat
^echo "1" "2" "3" | ^cat
```

The latter will output `1 2 3`. Hard to get that behaviour, so I thought new lines were a good compromise (just to give you where my mind was at when doing this).

We're in control of how this should behave with nushell though. What do you think?

Reply:
I've been wondering if we should fix internal echo so that it works closet to the external one. In that case they'd be on the same line

Reply:
> In that case they'd be on the same line

And would be less surprising to people. Probably not a bad idea.

I'll go ahead and assume that's how it works, and drop the `"\n"` 🙂